new-log-viewer/src/typings/states.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * Various states of the UI. UI components may be enabled or disabled based on the state.
+ */
+enum UI_STATE {
+
+    /**
+     * When there is no file opened in the viewer.
+     */
+    UNOPENED,
+
+    /**
+     * When a file load request is pending response.
+     */
+    FILE_LOADING,
+
+    /**
+     * When a fast request is pending response. In this state, UI elements are not visually
+     * disabled but instead ignore pointer events. Rapidly disabling/enabling UI elements is
+     * jarring to the user.
+     */
+    FAST_LOADING,
+
+    /**
+     * When a slow request is pending response.
+     */
+    SLOW_LOADING,
+
+    /**
+     * When the file is loaded.
+     */
+    READY,
+}
+
+/**
+ * Stateful elements in the UI.
+ */
+enum UI_ELEMENT {
+    PROGRESS_BAR,
+    NAVIGATION_BAR,
+    OPEN_FILE_BUTTON,
+    LOG_LEVEL_FILTER,
+    EXPORT_LOGS_BUTTON,
+    LOG_EVENT_NUM_DISPLAY,
+    DRAG_AND_DROP,
+}
+
+type UiElementRowType = {
+    [key in UI_ELEMENT]: boolean;
+};
+
+type UiStateGridType = {
+    [key in UI_STATE]: UiElementRowType;
+};
+
+/**
+ * Grid describing whether a `UI_ELEMENT` is enabled or disabled for every `UI_STATE`. Each
+ * `UI_STATE` has its own row with a boolean for each `UI_ELEMENT`. Boolean is `true` if the
+ * element is enabled, or `false` if disabled.
+ */
+const UI_STATE_GRID: UiStateGridType = Object.freeze({
+    [UI_STATE.UNOPENED]: {
+        [UI_ELEMENT.PROGRESS_BAR]: false,
+        [UI_ELEMENT.NAVIGATION_BAR]: false,
+        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
+        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
+        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
+        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
+        [UI_ELEMENT.DRAG_AND_DROP]: true,
```

Comment:
if alphabetizing those makes sense
```suggestion
        [UI_ELEMENT.DRAG_AND_DROP]: true,
        [UI_ELEMENT.EXPORT_LOGS_BUTTON]: false,
        [UI_ELEMENT.LOG_EVENT_NUM_DISPLAY]: false,
        [UI_ELEMENT.LOG_LEVEL_FILTER]: false,
        [UI_ELEMENT.NAVIGATION_BAR]: false,
        [UI_ELEMENT.OPEN_FILE_BUTTON]: true,
        [UI_ELEMENT.PROGRESS_BAR]: false,
```